master/internal/stream/metrics.go

```diff
@@ -0,0 +1,334 @@
+	ctx context.Context,
+	user model.User,
+	known string,
+	spec MetricSubscriptionSpec,
+) (
+	[]stream.PreparableMessage, error,
+) {
+	var out []stream.PreparableMessage
+
+	if len(spec.MetricIds) == 0 && len(spec.TrialIds) == 0 && len(spec.ExperimentIds) == 0 {
+		// empty subscription: everything known should be returned as deleted
+		out = append(out, stream.DeleteMsg{
+			Key:     MetricsDeleteKey,
+			Deleted: known,
+		})
+		return out, nil
+	}
+	// step 0: get user's permitted access scopes
+	accessMap, err := AuthZProvider.Get().GetMetricStreamableScopes(ctx, user)
+	if err != nil {
+		return nil, err
+	}
+	var accessScopes []model.AccessScopeID
+	for id, isPermitted := range accessMap {
+		if isPermitted {
+			accessScopes = append(accessScopes, id)
+		}
+	}
+
+	permFilter := func(q *bun.SelectQuery) *bun.SelectQuery {
+		if accessMap[model.GlobalAccessScopeID] {
+			return q
+		}
+		return q.Where("workspace_id in (?)", bun.In(accessScopes))
+	}
+
+	// step 1: calculate all ids matching this subscription
+	q := db.Bun().
+		NewSelect().
+		Table("metrics").
+		Column("metrics.id").
+		Join("JOIN trials t ON metrics.trial_id = t.id").
+		Join("JOIN experiments e ON t.experiment_id = e.id").
+		Join("JOIN projects p ON e.project_id = p.id").
+		OrderExpr("metrics.id ASC")
+	q = permFilter(q)
+
+	// Ignore mmf.Since, because we want appearances, which might not be have seq > spec.Since.
+	ws := stream.WhereSince{Since: 0}
+	if len(spec.MetricIds) > 0 {
+		ws.Include("metrics.id in (?)", bun.In(spec.MetricIds))
+	}
+	if len(spec.TrialIds) > 0 {
+		ws.Include("trial_id in (?)", bun.In(spec.TrialIds))
+	}
+	if len(spec.ExperimentIds) > 0 {
+		ws.Include("experiment_id in (?)", bun.In(spec.ExperimentIds))
+	}
+	q = ws.Apply(q)
+
+	var exist []int64
+	err = q.Scan(ctx, &exist)
+	if err != nil && errors.Cause(err) != sql.ErrNoRows {
+		log.Errorf("error: %v\n", err)
+		return nil, err
+	}
+
+	// step 2: figure out what was missing and what has appeared
+	missing, appeared, err := stream.ProcessKnown(known, exist)
+	if err != nil {
+		return nil, err
+	}
+
+	// step 3: hydrate appeared IDs into full MetricMsgs
+	var metricMsgs []*MetricMsg
+	if len(appeared) > 0 {
+		query := getMetricMsgsWithWorkspaceID(metricMsgs).
+			Where("metric_msg.id in (?)", bun.In(appeared))
+		query = permFilter(query)
+		err := query.Scan(ctx, &metricMsgs)
+		if err != nil && errors.Cause(err) != sql.ErrNoRows {
+			log.Errorf("error: %v\n", err)
+			return nil, err
+		}
+	}
+
+	// step 4: emit deletions and updates to the client
+	out = append(out, stream.DeleteMsg{
+		Key:     MetricsDeleteKey,
+		Deleted: missing,
+	})
+	for _, msg := range metricMsgs {
+		out = append(out, msg.UpsertMsg())
+	}
+	return out, nil
+}
+
+// MetricCollectSubscriptionModMsgs scrapes the database when a
+// user submits a new MetricSubscriptionSpec for initial matches.
+func MetricCollectSubscriptionModMsgs(ctx context.Context, addSpec MetricSubscriptionSpec) (
+	[]interface{}, error,
```

Comment:
nit (non-blocking) comment: just a reminder, but probably not important since I think this function is gonna get cut, that interface works but it should be `PreparedMessage` if it is to be functional

Reply:
Good call, I'll make sure to update the soft-resets PR to cut these functions out from the new streamable types!